cilium-health/launch/endpoint.go

```diff
@@ -102,10 +118,59 @@ func configureHealthRouting(netns, dev string, addressing *models.NodeAddressing
 	return err
 }
 
+// PingEndpoint attempts to make an API ping request to the local cilium-health
+// endpoint, and returns whether this was successful.
+func PingEndpoint() error {
+	if client == nil {
+		return fmt.Errorf("cilium-health endpoint hasn't yet been initialized")
+	}
+	_, err := client.Restapi.GetHello(nil)
+	return err
+}
+
+// CleanupEndpoint attempts to kill any existing cilium-health endpoint and
+// clean up its devices and pidfiles.
+func CleanupEndpoint(owner endpoint.Owner) {
+	pidfile := filepath.Join(option.Config.StateDir, healthPidfile)
+	if pid, err := ioutil.ReadFile(pidfile); err == nil {
+		// Existing cilium-health exists, kill it so we can create a
+		// new one and steal all its logs (om nom nom!)
+		pidInt, err := strconv.Atoi(string(pid))
+		if err == nil {
+			scopedLog := log.WithField("pid", pidInt)
+			scopedLog.Debug("Killing existing cilium-health instance")
+			if oldProc, err := os.FindProcess(pidInt); err == nil {
+				err = oldProc.Kill()
+				if err != nil {
+					scopedLog.WithError(err).Info("Couldn't kill cilium-health")
+				}
+			} else {
+				scopedLog.WithError(err).Debug("Couldn't find cilium-health")
+			}
+		}
+	}
+	// Ignore any errors; we just need to remove the old pidfile.
+	_ = os.RemoveAll(pidfile)
+
+	// Remove device
+	scopedLog := log.WithField(logfields.Veth, vethName)
+	if link, err := netlink.LinkByName(vethName); err == nil {
+		err = netlink.LinkDel(link)
+		if err != nil {
+			scopedLog.WithError(err).Info("Couldn't delete cilium-health device")
+		}
+	} else {
+		scopedLog.WithError(err).Debug("Didn't find existing device")
+	}
+}
+
 // LaunchAsEndpoint launches the cilium-health agent in a nested network
 // namespace and attaches it to Cilium the same way as any other endpoint,
 // but with special reserved labels.
-func LaunchAsEndpoint(owner endpoint.Owner, hostAddressing *models.NodeAddressing, opts *option.BoolOptions) context.CancelFunc {
+func LaunchAsEndpoint(owner endpoint.Owner, hostAddressing *models.NodeAddressing) error {
+
+	CleanupEndpoint(owner)
```

Comment:
This is an unexpected side-effect. Since we only call `LaunchAsEndpoint` in a few places, we can always just directly call `CleanupEndpoint` there. Otherwise, if you're really worried about one day launching without cleanup could you update the comment, at least.